### PR TITLE
Enable noUncheckedIndexedAccess for merge tree

### PR DESCRIPTION
**Pull request by RishhiB**
Enabling noUncheckedIndexedAccess for merge tree
We are enabling `noUncheckedIndexedAccess` to protect us from changing a type of a record where there was a named property and reduce the risk of runtime errors by ensuring that indexed access on objects are properly checked for undefined values.

Arrays are impacted by this and must be annotated even though it is not the reason for using the setting

[AB#8216](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8216)

### CODE REVIEW

File: packages/dds/merge-tree/src/properties.ts [hunk 1]
```diff
@@ -87,7 +88,8 @@ export function clone<T>(extension: MapLike<T> | undefined) {
 	// eslint-disable-next-line guard-for-in, no-restricted-syntax
 	for (const key in extension) {
 		const v = extension[key];
-		if (v !== null) {
+		// TODO Non null asserting, why is this not null?
+		if (v !== null && v !== undefined) {
```
Comment (Abe27342) [inline packages/dds/merge-tree/src/properties.ts:92]: i don't like including behavioral changes as part of this PR except where they are statically analyzable. Changing the behavior of clone/extends here might have unexpected consequences in other bits of merge tree.

File: packages/dds/merge-tree/src/properties.ts [hunk 2]
```diff
@@ -63,7 +63,8 @@ export function extend<T>(base: MapLike<T>, extension: MapLike<T> | undefined) {
 		// eslint-disable-next-line guard-for-in, no-restricted-syntax
 		for (const key in extension) {
 			const v = extension[key];
-			if (v === null) {
+			// TODO Non null asserting, why is this not null?
+			if (v === null || v === undefined) {
```
Comment (Abe27342) [inline packages/dds/merge-tree/src/properties.ts:67]: Similar to my feedback on clone, I'd avoid making this change.

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 3]
```diff
@@ -594,7 +595,7 @@ export class MergeTree {
 				this.blockUpdate(block);
 			}
 
-			return blocks.length === 1 // If there is only one block at this layer...
+			return blocks.length === 1 && blocks[0] !== undefined // If there is only one block at this layer...
```
Comment (Abe27342) [inline packages/dds/merge-tree/src/mergeTree.ts:598]: another behavioral change

Comment (Abe27342) [reply packages/dds/merge-tree/src/mergeTree.ts:598]: well, not really. but the check is redundant.

File: packages/dds/merge-tree/src/properties.ts [hunk 4]
```diff
@@ -88,7 +91,9 @@ export function clone<T>(extension: MapLike<T> | undefined) {
 	for (const key in extension) {
 		const v = extension[key];
 		if (v !== null) {
-			cloneMap[key] = v;
+			// TODO why are we non null asserting here?
```
Comment (Abe27342) [inline packages/dds/merge-tree/src/properties.ts:94]:
```suggestion
			// If `v` is undefined, undefined must have been assignable to `T`.
```

**Review by Abe27342 — COMMENTED** [review]
looks good, ping me once you've published it and I can approve.

**Comment by RishhiB** [comment]
I am assuming I will go back to all the TODO comments and update them with the standardized comment

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**